### PR TITLE
Fixed : Split expenses should be added to the report if the report is selected in an expense form

### PR DESCRIPTION
**Pull request by siddharth0-0**
allgod

somme

image

image

### CODE REVIEW

File: src/app/fyle/split-expense/split-expense.page.ts [hunk 1]
```diff
@@ -237,12 +241,28 @@ export class SplitExpensePage implements OnInit {
...
     return forkJoin(splitExpense$).pipe(
       switchMap((data: any) => {
+        this.saveToReports(data);
         const txnIds = data.txns.map((txn) => txn.id);
         return this.splitExpenseService.linkTxnWithFiles(data).pipe(map(() => txnIds));
       })
     );
   }
 
+  saveToReports(data) {
```
Comment (Dimple16) [inline src/app/fyle/split-expense/split-expense.page.ts:251]: `addToReport`

File: src/app/fyle/split-expense/split-expense.page.ts [hunk 2]
```diff
@@ -237,12 +241,28 @@ export class SplitExpensePage implements OnInit {
 
     return forkJoin(splitExpense$).pipe(
       switchMap((data: any) => {
+        this.saveToReports(data);
```
Comment (Dimple16) [inline src/app/fyle/split-expense/split-expense.page.ts:244]: shouldn't we call this only when report ID is present?

File: src/app/fyle/split-expense/split-expense.page.ts [hunk 3]
```diff
@@ -237,12 +241,28 @@ export class SplitExpensePage implements OnInit {
...
+    if (reportId) {
+      const splitExpense = data.txns.map((txn) => txn);
+      const txnIds = splitExpense
+        .filter((tx) => {
+          if (tx.state === 'COMPLETE') {
+            return true;
+          }
+        })
+        .map((txn) => txn.id);
+      return this.reportService.addTransactions(reportId, txnIds).subscribe(noop);
```
Comment (Dimple16) [inline src/app/fyle/split-expense/split-expense.page.ts:262]: let us add a toast message here as to how many expenses got successfully added to report

**Review by Dimple16 — CHANGES_REQUESTED** [review]
few comments

File: src/app/fyle/split-expense/split-expense.page.ts [hunk 4]
```diff
@@ -64,6 +65,12 @@ export class SplitExpensePage implements OnInit {
 
   showErrorBlock: boolean;
 
+  reportId: string;
+
+  splitExpenseTxn: any;
```
Comment (Dimple16) [inline src/app/fyle/split-expense/split-expense.page.ts:70]: let us avoid using `any`

File: src/app/fyle/split-expense/split-expense.page.ts [hunk 5]
```diff
@@ -243,12 +250,80 @@ export class SplitExpensePage implements OnInit {
...
+        const expensesAddedToReportSnackBar = this.matSnackBar.openFromComponent(ToastMessageComponent, {
+          ...this.snackbarProperties.setSnackbarProperties('success', toastMessageData),
+          panelClass: ['msb-success-with-camera-icon'],
+        });
+
+        expensesAddedToReportSnackBar.onAction().subscribe(() => {
+          this.router.navigate(['/', 'enterprise', 'my_view_report', { id: this.reportId, navigateBack: true }]);
+        });
+      } else {
+        const message = 'Your expense was split successfully. But no Expense was added to report.';
```
Comment (Dimple16) [inline src/app/fyle/split-expense/split-expense.page.ts:311]: I guess showing a green toast here would mislead, can we show the dark blue toast for this case?

**Review by Dimple16 — CHANGES_REQUESTED** [review]
few comments

File: src/app/fyle/split-expense/split-expense.page.ts [hunk 6]
```diff
@@ -64,6 +65,12 @@ export class SplitExpensePage implements OnInit {
 
   showErrorBlock: boolean;
 
+  reportId: string;
+
+  splitExpenseTxn;
+
+  completeTxnIds;
```
Comment (Dimple16) [inline src/app/fyle/split-expense/split-expense.page.ts:72]: is this a list?

**Review by Dimple16 — APPROVED** [review]

File: src/app/fyle/split-expense/split-expense.page.ts [hunk 7]
```diff
@@ -243,12 +251,84 @@ export class SplitExpensePage implements OnInit {
...
+          return true;
+        }
+      })
+      .map((txn) => txn.id);
+    return this.reportService.addTransactions(reportId, this.completeTxnIds).subscribe(noop);
+  }
+
+  showSuccessToast() {
+    if (this.reportId) {
+      if (this.completeTxnIds.length === this.splitExpenseTxn.length) {
```
Comment (Julias0) [inline src/app/fyle/split-expense/split-expense.page.ts:278]: the structure of each if condition is duplicated, try to cut down the duplication here and make this simpler too.

File: src/app/fyle/split-expense/split-expense.page.ts [hunk 8]
```diff
@@ -243,12 +251,84 @@ export class SplitExpensePage implements OnInit {
...
+    const reportId = this.reportId;
+    this.splitExpenseTxn = data.txns.map((txn) => txn);
+    this.completeTxnIds = this.splitExpenseTxn
+      .filter((tx) => {
+        if (tx.state === 'COMPLETE') {
+          return true;
+        }
+      })
+      .map((txn) => txn.id);
+    return this.reportService.addTransactions(reportId, this.completeTxnIds).subscribe(noop);
```
Comment (Julias0) [inline src/app/fyle/split-expense/split-expense.page.ts:273]: dont subscribe to observables inside like this. Make this a separate switch map

switchMap(data => {
  if (this.reportId) {
    return this.addToReport(data).pipe(map(_=> data))
  } else {
    return of(data);
  }
})

File: src/app/fyle/split-expense/split-expense.page.ts [hunk 9]
```diff
@@ -242,13 +250,78 @@ export class SplitExpensePage implements OnInit {
     }
 
     return forkJoin(splitExpense$).pipe(
+      switchMap((data: any) => {
+        this.splitExpenseTxn = data.txns.map((txn) => txn);
+        this.completeTxnIds = this.splitExpenseTxn
+          .filter((tx) => {
```
Comment (Julias0) [inline src/app/fyle/split-expense/split-expense.page.ts:256]: .filter((tx) => tx.state === 'COMPLETE')

File: src/app/fyle/split-expense/split-expense.page.ts [hunk 10]
```diff
@@ -242,13 +250,78 @@ export class SplitExpensePage implements OnInit {
...
+        }
+      }),
       switchMap((data: any) => {
         const txnIds = data.txns.map((txn) => txn.id);
         return this.splitExpenseService.linkTxnWithFiles(data).pipe(map(() => txnIds));
       })
     );
   }
 
+  toastWithCTA(toastMessage) {
```
Comment (Julias0) [inline src/app/fyle/split-expense/split-expense.page.ts:275]: you can combine this and the below one by adding a boolean - withCTA: boolean parameter

**Review by Julias0 — APPROVED** [review]

**Pull request merged**